modules/login.go

```diff
@@ -59,15 +67,39 @@ func (l *LoginReq) Login(ctx context.Context) (map[string]string, error) {
 	}
 
 	// id is UUID(16 bytes) Google's UUID.
+	// Query user and get password hash
 	var id uuid.UUID
 	var role string
+	var storedPasswordHash string
+	err = db.QueryRow(ctx, "SELECT id, role, password FROM users WHERE username = $1 OR email = $2", l.UserName, l.Email).Scan(&id, &role, &storedPasswordHash)
 
-	err = db.QueryRow(ctx, "SELECT id, role FROM users WHERE (username = $1 OR email = $2) AND password = $3", l.UserName, l.Email, l.Password).Scan(&id, &role)
 	if err != nil {
 		logger.Error(fmt.Sprintf("Login: failed to query user: %v", err), err)
 		return nil, fmt.Errorf("invalid username/email or password")
 	}
 
+	if isBcryptHash(storedPasswordHash) {
+		err = bcrypt.CompareHashAndPassword([]byte(storedPasswordHash), []byte(l.Password))
+		if err != nil {
+			logger.Info("Login: invalid password attempt")
+			return nil, fmt.Errorf("invalid username/email or password")
+		}
+	} else {
+		// Compare plain text for legacy users
+		if storedPasswordHash != l.Password {
+			logger.Info("Login: invalid password attempt (legacy)")
+			return nil, fmt.Errorf("invalid username/email or password")
+		}
+
+		// Rehash and update DB for this user
+		newHash, err := bcrypt.GenerateFromPassword([]byte(l.Password), bcrypt.DefaultCost)
+		if err == nil {
+			_, _ = db.Exec(ctx, "UPDATE users SET password = $1 WHERE id = $2", string(newHash), id)
+			logger.Info(fmt.Sprintf("Upgraded password hash for user %v", id))
+		}
+	}
```

Comment:
it's great that you thought about this. But the usual practice is where you write a migration script and add it as part of the initialization of the service. In that migration script you migrate all passwords to the hashed form. Your approach fails in cases where users have set actual passwords starting with $2b ... etc